### ISSUE TITLE
Pre-Prerelease Card Viewer Brainstorming

### ISSUE THREAD

**Issue by mrphlip** [mrphlip/lrrbot#160]
So, the pre-prerelease is going to be a paper event, which I assume means we're going to want to have the card viewer working. It was mentioned in passing during the stream that Paul is trying to get it organised that the new cards will be in the OCR, but if not they'll fall back to a manual process to put the cards up on-screen. Which made me think about how this is going to work in chat.

It's the day after the full spoiler is released, which usually means that Gatherer likely won't be updated yet (usually that happens a couple days later). MTGJSON definitely won't be updated yet. We can get around that, there are other places we can get the card text (scrape Mythic Spoiler or Salvation, or if worse comes to worst, just type them all in, there's only a couple hundred of them).

But that still leaves up the question of communication - if Gatherer isn't out yet, we don't know the `multiverseid`s, and that's how we currently get the cards across. If they start sending links to the images from the Card Image Gallery over the pubnub channel, the bot isn't going to recognise those. _Making_ it recognise those URLs wouldn't be an impossible job either, but we'd need to know that's the plan.

I guess the question is... what's the plan here from the LRR side, so we know what we have to work with? How much has been figured out yet, and how much is still to-be-decided?

**Comment by mrphlip** [comment]
Attn @paul-lrr

**Comment by paul-lrr** [comment]
The current status on the card recognizer for the SOI event is still in flux. I'm in communication with several companies that make card recognition software and we are trying to figure out if it will be possible. Since the cards won't be in gatherer and probably won't have multiverse ids (not sure if they are predictable) we are probably going to be using the collector number. WotC is going to be supplying us with high res images of all the cards so we can have them all locally instead of relying on the spoiler image gallery. 

The tl;dr is that regardless of whether we use a recognizer or do it manually, you should probably expect that lrrbot will be getting the collector number of the cards instead of the multiverse id

**Comment by mrphlip** [comment]
Collector numbers would work. Perhaps, for future-proofing, have it send `SOI-123` or somesuch. As a separate channel to the `xsplit_image` one. Then it should be easier if you do this again next set when there's both SOI and EMN cards to deal with.

Looking at the double-faced cards, it looks like they have the same collector's number on both sides... but we could take a cue from the last time, where they had "a" and "b" appended... so, like, Huntmaster of the Fells is `DKA-140a` and Ravager of the Fells is `DKA-140b`... so we could do the same thing and make Archangel Avacyn `SOI-5a` and Avacyn, the Purifier `SOI-5b`, even though they're both printed as collector number "005/297" on the new frames.

**Comment by andreasots** [comment]
> As a separate channel to the `xsplit_image` one.

We are on a different channel: `lrrbot`.

As for what to send over the channel how about something like `{"multiverseid": 262875}` or `{"number:" "140a", "set": "DKA"}` or `{"name": "Huntmaster of the Fells"}` for Huntmaster of the Fells? If not in JSON then in whatever encoding Paul finds convenient.

**Comment by mrphlip** [comment]
OK, I have a script that'll scrape the cards off MTGSalvation... modulo a few typos and formatting gripes, it's just as good as the real thing. And we have the capability to look cards up by set and collector number (for any set), so once we've settled on a format, that can slot right in alongside multiverseid lookups.

**Comment by paul-lrr** [comment]
Ok, looks like we have a solution for automated card recognition for the SOI event. The guys who make Decked Builder Card Cam have actually built a custom piece of software for us, so that is pretty cool. 

In order to keep things as simple as possible on my end, I would like to continue with what I was doing for LRRbot before and just send the same image url that we are using for the overlay. For the normal sets, that is the Gatherer url (`http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=<<ID>>&type=card`) but for the SOI event, where the images are not in Gatherer yet, it will be a reference to an image on the local computer. It will be in the form `http://localhost/cards/<<SET>>-<<COLLECTOR NUM>>.jpg` so for a SOI card, it will be `http://localhost/cards/SOI-123.jpg`. I like the suggestion from mrphlip for how to handle double face cards, so let's have the initial side be `a` and the flip side be `b`. So **Archangel Avacyn** will come in as `http://localhost/cards/SOI-5a.jpg` and **Avacyn, the Purifier** will come in as `http://localhost/cards/SOI-5b.jpg`. Does that all make sense? I figure it will be pretty straight forward for you guys to extract the set code and collector number from the url and use that for find the card text in your system.

Once the cards are in Gatherer, the DFCs will have multiverse ids as normal (interesting note, the two sides of DFCs have _different_ multiverse ids that are not necessarily even consecutive).

**Comment by andreasots** [comment]
Done.

**Comment by andreasots** [comment]
I looked at the logs and found entries like
```
[2016-03-21 16:59:13,092] ERROR:cardviewer:Failed to extract set and collector number from 'http://localhost/cards/ALA_212.jpg'
```

Now that format is recognised as well.

**Comment by paul-lrr** [comment]
Oops! Sorry about that. I was testing and didn't realize it was sending to lrrbot. (This is why having it need to be turned on before it shows up in the chat was a good idea ^_^' )

Obviously the underscore vs dash doesn't really make any practical difference (as long as we choose one and stick with it), but I think I will go with the dash in the future.

**Comment by paul-lrr** [comment]
I am hoping to be able to load the tokens into the scan database as well. Since tokens have their own collector numbers, I am adding `t` to the filename to separate it from the card with that number. 

eg. The Angel token would be `SOI-001t.png`

**Comment by mrphlip** [comment]
> Also, as you can see from that example, I figured I would 0 pad the numbers to match how they are printed on the cards. Is that a problem?

Not a _problem_, we'd just have to handle it (since the data source doesn't have leading zeroes, and we're doing string matching not number matching, because of the `a`s and `b`s and whatnot).

As for tokens... we certainly _could_. I don't know of any databases for that, though, so it'd only have the sets we manually enter... Wouldn't take too long, though, SOI has, what, 18 of them?

I'd have to do some amount of cleverness to make them not show up in `!card` search results (or, alternatively, some clever way of making them searchable) because different sets can have tokens with the same name and different stats. Maybe they can be listed as "Angel (SOI Token)" or whatnot.

I'm not completely convinced it's worth the effort, though... Of those 18, there's, what... 2 emblems that are unlikely to come up (and if they do, you can get the same effect by just `!card`ing the planeswalker), 6 clue tokens (that people will quickly learn how they work, it's a keyword mechanic after all), and 9 vanilla or french vanilla creatures (which aren't really worth spilling into chat). Basically the only interesting one that people might need to actually _read_ is the Devil token.

Which is not to say you can't still just scan them and put them on the overlay, that would be pretty cool (if only to show off the art), I just don't think it's worth making it print "Insect | Creature - Insect [1/1]" in the chat when you do so.

**Comment by andreasots** [comment]
Leading zeroes now stripped from collector number.

**Comment by paul-lrr** [comment]
good point about the tokens. I guess just be aware that ids ending in `t` don't need to be shown in chat. 

Frankly the leading zero thing was more of an aesthetic choice then any real logic. I can easily remove them if you want.

**Comment by andreasots** [comment]
The collector numbers on the scraped SOI cards need to be double-checked before the event.

```
18:20 < qrpth> !card Arlinn Kord
18:20 < lrrbot> Arlinn Kord [2RG] (back: The Gitrog Monster) | Planeswalker — Arlinn [3] 
                | +1: Until end of turn, up to one target creature gets +2/+2 and gains 
                vigilance and haste. / 0: Put a 2/2 green Wolf creature token onto the 
                battlefield. Transform Arlinn Kord.
18:21 < qrpth> !card The Gitrog Monster
18:21 < lrrbot> The Gitrog Monster [3BG] (front: Arlinn Kord) | Legendary Creature — Frog 
                Horror [6/6] | Deathtouch / At the beginning of your upkeep, sacrifice 
                The Gitrog Monster unless you sacrifice a land. / You may play an 
                additional land on each of your turns. / Whenever one or more land cards 
                are put into your graveyard from anywhere, draw a card.
```

[Arlinn Kord](http://www.mtgsalvation.com/cards/shadows-over-innistrad/26750-arlinn-kord)'s number is 243 but on MTG Salvation it's 245, the same [The Gitrog Monster](http://www.mtgsalvation.com/cards/shadows-over-innistrad/26840-the-gitrog-monster).

**Comment by andreasots** [comment]
And now Skeleton Key as well:
```
19:02 < qrpth> !card skeleton key
19:02 < lrrbot> Skeleton Key [1] (front: Arlinn Kord) | Artifact — Equipment | Equipped 
                creature has Skulk / Whenever equipped creature deals damage to a player, 
                you may draw a card. If you do discard a card. / Equip {2}
```

**Comment by andreasots** [comment]
[All cards with correct info](https://gist.github.com/andreasots/ec565482c7413e993504). It's getting late so I might have missed something. That version is in LRRbot so everything should be ready.

**Comment by mrphlip** [comment]
Oh... I missed your comment, and I'm about halfway through proofreading all the cards myself.

I guess I'll finish, then compare mine to yours. Hopefully anything I missed, you caught, and vice versa.

There's quite a few errors in the Salvation cards...

**Comment by mrphlip** [comment]
OK, I've merged my changes and your changes (for the most part by just shoving both into the repository and letting git sort it out), and have loaded that into lrrbot.

**Comment by paul-lrr** [comment]
Awesome! Thanks to all of you for working this out. If we get to do an event like this again (fingers crossed!) hopefully we can refine the process to make require less manual data input.

Also, just as an FYI, it may be occasionally necessary for myself or Ian (who will also be running tech) to manually override the card being displayed on the overlay with another card image stored locally on the stream computer. If this happens, lrrbot will not receive any notification because the image data (not the filename) is being directly sent to the overlay. I don't think this will need to happen very often, if at all, but I just wanted to let you know in case lrrbot seems to miss a card.

**Comment by andreasots** [comment]
Everything is working perfectly.
